DataFormats/Histograms/interface/MonitorElementCollection.h

```diff
@@ -0,0 +1,229 @@
+    std::string objname_;
+
+  public:
+    enum class Type { DIR, DIR_AND_NAME };
+
+    std::string const& getDirname() const { return dirname_; }
+    std::string const& getObjectname() const { return objname_; }
+
+    // Clean up the path and normalize it to preserve certain invariants.
+    // Instead of reasoning about whatever properties of paths, we just parse
+    // the thing and build a normalized instance with no slash in the beginning
+    // and a slash in the end.
+    // Type of string `path` could be just directory name, or
+    // directory name followed by the name of the monitor element
+    void set(std::string path, Path::Type type) {
+      std::string in(path);
+      std::vector<std::string> buf;
+      std::regex dir("^/*([^/]+)");
+      std::smatch m;
+
+      while (std::regex_search(in, m, dir)) {
+        if (m[1] == "..") {
+          if (!buf.empty()) {
+            buf.pop_back();
+          }
+        } else {
+          buf.push_back(m[1]);
+        }
+        in = m.suffix().str();
+      }
+
+      // Construct dirname_ and object_name
+      dirname_ = "";
+      objname_ = "";
+      int numberOfItems = buf.size();
+      for (int i = 0; i < numberOfItems; i++) {
+        if (i == numberOfItems - 1) {
+          // Processing last component...
+          if (type == Path::Type::DIR_AND_NAME) {
+            objname_ = buf[i];
+          } else if (type == Path::Type::DIR) {
+            dirname_ += buf[i] + "/";
+          }
+        } else {
+          dirname_ += buf[i] + "/";
+        }
+      }
+    }
+
+    bool operator==(Path const& other) const {
+      return this->dirname_ == other.dirname_ && this->objname_ == other.objname_;
+    }
+  };
+
+  // Metadata about the ME. The range is included here in case we have e.g.
+  // multiple per-lumi histograms in one collection. For a logical comparison,
+  // one should look only at the name.
+  struct Key {
+    Path path_;
+
+    // Run number (and optionally lumi number) that the ME belongs to.
+    edm::LuminosityBlockID id_;
+    Scope scope_;
+    Kind kind_;
+
+    bool operator<(Key const& other) const {
+      auto makeKeyTuple = [](Key const& k) {
+        return std::make_tuple(
+            k.path_.getDirname(), k.path_.getObjectname(), k.scope_, k.id_.run(), k.id_.luminosityBlock());
+      };
+
+      return makeKeyTuple(*this) < makeKeyTuple(other);
+    }
+  };
+
+  bool operator<(MonitorElementData const& other) const { return this->key_ < other.key_; }
+
+  // The only non class/struct members
+  Key key_;
+  Value value_;
+};
+
+// For now, no additional (meta-)data is needed apart from the MEs themselves.
+// The framework will take care of tracking the plugin and LS/run that the MEs
+// belong to.
+
+// Only to hold the mergeProduct placeholder for now.
+class MonitorElementCollection {
+  std::vector<std::unique_ptr<const MonitorElementData>> data_;
+
+public:
+  void push_back(std::unique_ptr<const MonitorElementData>&& value) {
```

Comment:
I think it is cleaner to do `push_back(std::unique_ptr<const MonitorElementData> value)` as that will match what other standard containers of `std::unique_ptr` will do.